### PR TITLE
refact(cstor): added last update fields for cstor objects

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

**What this PR does / why we need it**:
This PR adds the `LastUpdatedTime` and `LastTransitionTime` field for the status of `cstorvolume`, `cstorvolumereplica` and `cstorpool`.

sample csp :
```
apiVersion: openebs.io/v1alpha1
  kind: CStorPool
  metadata:
    annotations:
      openebs.io/csp-lease: '{"holder":"openebs/cstor-sparse-pool-8y6d-7fb44586c6-xfphf","leaderTransition":1}'
    creationTimestamp: "2019-07-03T06:51:17Z"
    generation: 60
    labels:
      kubernetes.io/hostname: minikube
      openebs.io/cas-template-name: cstor-pool-create-default-1.1.0
      openebs.io/cas-type: cstor
      openebs.io/storage-pool-claim: cstor-sparse-pool
      openebs.io/version: 1.1.0
    name: cstor-sparse-pool-8y6d
    ownerReferences:
    - apiVersion: openebs.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: StoragePoolClaim
      name: cstor-sparse-pool
      uid: f4340f70-9d5e-11e9-907e-54e1ad5e8320
    resourceVersion: "8899"
    selfLink: /apis/openebs.io/v1alpha1/cstorpools/cstor-sparse-pool-8y6d
    uid: f52e9cc8-9d5e-11e9-907e-54e1ad5e8320
  spec:
    group:
    - blockDevice:
      - deviceID: /var/openebs/sparse/1-ndm-sparse.img
        inUseByPool: true
        name: sparse-089efd60dca9408ff135f4a2f5ebb545
    poolSpec:
      cacheFile: ""
      overProvisioning: false
      poolType: striped
  status:
    capacity:
      free: 9.94G
      total: 9.94G
      used: 1.64M
    lastTransitionTime: "2019-07-03T06:51:33Z"
    lastUpdateTime: "2019-07-03T07:20:00Z"
    phase: Healthy
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
sample cvr:
```
apiVersion: openebs.io/v1alpha1
  kind: CStorVolumeReplica
  metadata:
    annotations:
      cstorpool.openebs.io/hostname: minikube
      isRestoreVol: "false"
      openebs.io/storage-class-ref: |
        name: openebs-cstor-sparse
        resourceVersion: 974
    creationTimestamp: "2019-07-03T07:43:07Z"
    finalizers:
    - cstorvolumereplica.openebs.io/finalizer
    generation: 9
    labels:
      cstorpool.openebs.io/name: cstor-sparse-pool-8y6d
      cstorpool.openebs.io/uid: f52e9cc8-9d5e-11e9-907e-54e1ad5e8320
      cstorvolume.openebs.io/name: pvc-31799907-9d66-11e9-885e-54e1ad5e8320
      openebs.io/cas-template-name: cstor-volume-create-default-1.1.0
      openebs.io/persistent-volume: pvc-31799907-9d66-11e9-885e-54e1ad5e8320
      openebs.io/version: 1.1.0
    name: pvc-31799907-9d66-11e9-885e-54e1ad5e8320-cstor-sparse-pool-8y6d
    namespace: openebs
    resourceVersion: "11516"
    selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/cstorvolumereplicas/pvc-31799907-9d66-11e9-885e-54e1ad5e8320-cstor-sparse-pool-8y6d
    uid: 32f0f48c-9d66-11e9-885e-54e1ad5e8320
  spec:
    capacity: 4G
    targetIP: 10.97.224.174
    zvolWorkers: ""
  status:
    capacity:
      totalAllocated: 6K
      used: 6K
    lastTransitionTime: "2019-07-03T07:43:55Z"
    lastUpdateTime: "2019-07-03T07:45:57Z"
    phase: Healthy
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

sample cv:
```
apiVersion: openebs.io/v1alpha1
  kind: CStorVolume
  metadata:
    annotations:
      openebs.io/fs-type: ext4
      openebs.io/lun: "0"
      openebs.io/storage-class-ref: |
        name: openebs-cstor-sparse
        resourceVersion: 974
    creationTimestamp: "2019-07-03T07:43:07Z"
    generation: 6
    labels:
      openebs.io/cas-template-name: cstor-volume-create-default-1.1.0
      openebs.io/persistent-volume: pvc-31799907-9d66-11e9-885e-54e1ad5e8320
      openebs.io/version: 1.1.0
    name: pvc-31799907-9d66-11e9-885e-54e1ad5e8320
    namespace: openebs
    resourceVersion: "11499"
    selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/cstorvolumes/pvc-31799907-9d66-11e9-885e-54e1ad5e8320
    uid: 32cff794-9d66-11e9-885e-54e1ad5e8320
  spec:
    capacity: 4G
    consistencyFactor: 1
    iqn: iqn.2016-09.com.openebs.cstor:pvc-31799907-9d66-11e9-885e-54e1ad5e8320
    nodeBase: iqn.2016-09.com.openebs.cstor
    replicationFactor: 1
    status: Init
    targetIP: 10.97.224.174
    targetPort: "3260"
    targetPortal: 10.97.224.174:3260
  status:
    lastTransitionTime: "2019-07-03T07:43:51Z"
    lastUpdateTime: "2019-07-03T07:45:51Z"
    phase: Healthy
    replicaStatuses:
    - checkpointedIOSeq: "0"
      inflightRead: "0"
      inflightSync: "0"
      inflightWrite: "0"
      mode: Healthy
      quorum: "1"
      replicaId: "753706992609146025"
      upTime: 142
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: cmd/cstor-pool-mgmt/controller/pool-controller/handler.go [hunk 1]
```diff
@@ -70,6 +70,10 @@ func (c *CStorPoolController) syncHandler(key string, operation common.QueueOper
 		glog.Warning("Empty status recieved for csp status in sync handler")
 		return nil
 	}
+	if cspObject.Status.Phase != apis.CStorPoolPhase(status) {
+		cspObject.Status.LastTransitionTime = metav1.Now()
```
Comment (AmitKumarDas) [inline cmd/cstor-pool-mgmt/controller/pool-controller/handler.go]: can we invoke metav1.Now() only once?
```go
cspObject.Status.LastUpdateTime = metav1.Now()
if cspObject.Status.Phase != apis.CStorPoolPhase(status) {
  cspObject.Status.Phase = apis.CStorPoolPhase(status)
  cspObject.Status.LastTransitionTime = cspObject.Status.LastUpdateTime
}
```

File: cmd/cstor-pool-mgmt/controller/replica-controller/handler.go [hunk 2]
```diff
@@ -80,7 +83,10 @@ func (c *CStorVolumeReplicaController) syncHandler(
 		// status holds more importance than error
 		return nil
 	}
-
+	if cvrGot.Status.Phase != apis.CStorVolumeReplicaPhase(status) {
+		cvrGot.Status.LastTransitionTime = metav1.Now()
+	}
+	cvrGot.Status.LastUpdateTime = metav1.Now()
```
Comment (AmitKumarDas) [inline cmd/cstor-pool-mgmt/controller/replica-controller/handler.go]: Look at above comment

File: cmd/cstor-pool-mgmt/controller/pool-controller/handler.go (hunk 1 above)
Comment (shubham14bajpai) [reply cmd/cstor-pool-mgmt/controller/pool-controller/handler.go]: Done

File: cmd/cstor-pool-mgmt/controller/replica-controller/handler.go (hunk 2 above)
Comment (shubham14bajpai) [reply cmd/cstor-pool-mgmt/controller/replica-controller/handler.go]: Done

File: cmd/cstor-volume-mgmt/controller/volume-controller/handler.go [hunk 3]
```diff
@@ -110,16 +139,27 @@ func (c *CStorVolumeController) cStorVolumeEventHandler(operation common.QueueOp
...
 			// if replicas are zero set the status as init
 			if len(volStatus.ReplicaStatuses) == 0 {
-				cStorVolumeGot.Status.Phase = apis.CStorVolumePhase(common.CVStatusInit)
+				cStorVolumeGot.Status.Phase = apis.CStorVolumePhase(
+					common.CVStatusInit,
+				)
 			}
 		}
+		cStorVolumeGot.Status.LastUpdateTime = metav1.Now()
+		if cStorVolumeGot.Status.Phase != lastKnownPhase {
```
Comment (mittachaitu) [inline cmd/cstor-volume-mgmt/controller/volume-controller/handler.go]: IMO we need to update `LastTransitionTime` even if there is change any one of the field under `status` section. In this PR we are updating only when there is phase change but `LastTransitionTime` is meant for status.
One way to do this by using `reflect.DeepEqual`

What would you say @vishnuitta @AmitKumarDas

Comment (mittachaitu) [reply cmd/cstor-volume-mgmt/controller/volume-controller/handler.go]: I think same thing need to do on `cStorPool and cvr`

Comment (mittachaitu) [reply cmd/cstor-volume-mgmt/controller/volume-controller/handler.go]: As we discussed `LastTransitionTime` is only for particular phase of the resource

File: cmd/cstor-volume-mgmt/controller/volume-controller/handler.go [hunk 4]
```diff
@@ -18,29 +18,33 @@ package volumecontroller
...
+
 	"strings"
 
 	"github.com/golang/glog"
 	"github.com/openebs/maya/cmd/cstor-volume-mgmt/controller/common"
 	"github.com/openebs/maya/cmd/cstor-volume-mgmt/volume"
 	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
 	"github.com/openebs/maya/pkg/client/k8s"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (shubham14bajpai) [reply cmd/cstor-volume-mgmt/controller/volume-controller/handler.go]: Done

File: cmd/cstor-pool-mgmt/controller/pool-controller/handler.go [hunk 5]
```diff
@@ -70,10 +70,10 @@ func (c *CStorPoolController) syncHandler(key string, operation common.QueueOper
 		glog.Warning("Empty status recieved for csp status in sync handler")
 		return nil
 	}
+	cspObject.Status.LastUpdateTime = metav1.Now()
 	if cspObject.Status.Phase != apis.CStorPoolPhase(status) {
-		cspObject.Status.LastTransitionTime = metav1.Now()
+		cspObject.Status.LastTransitionTime = cspObject.Status.LastUpdateTime
 	}
-	cspObject.Status.LastUpdateTime = metav1.Now()
 	cspObject.Status.Phase = apis.CStorPoolPhase(status)
```
[the remark this answers is not in the document]
Comment (shubham14bajpai) [reply cmd/cstor-pool-mgmt/controller/pool-controller/handler.go]: Done

File: cmd/cstor-pool-mgmt/controller/replica-controller/handler.go [hunk 6]
```diff
@@ -83,10 +83,10 @@ func (c *CStorVolumeReplicaController) syncHandler(
...
 		return nil
 	}
+	cvrGot.Status.LastUpdateTime = metav1.Now()
 	if cvrGot.Status.Phase != apis.CStorVolumeReplicaPhase(status) {
-		cvrGot.Status.LastTransitionTime = metav1.Now()
+		cvrGot.Status.LastTransitionTime = cvrGot.Status.LastUpdateTime
 	}
-	cvrGot.Status.LastUpdateTime = metav1.Now()
 	// set phase based on received status
 	cvrGot.Status.Phase = apis.CStorVolumeReplicaPhase(status)
```
[the remark this answers is not in the document]
Comment (shubham14bajpai) [reply cmd/cstor-pool-mgmt/controller/replica-controller/handler.go]: Done

File: cmd/cstor-pool-mgmt/controller/pool-controller/handler.go [hunk 7]
```diff
@@ -57,7 +57,7 @@ func (c *CStorPoolController) syncHandler(key string, operation common.QueueOper
 	csp, err := newCspLease.Hold()
 	cspObject, ok := csp.(*apis.CStorPool)
 	if !ok {
-		fmt.Errorf("expected csp object but got %#v", cspObject)
+		return fmt.Errorf("expected csp object but got %#v", cspObject)
```
Comment (vishnuitta) [inline cmd/cstor-pool-mgmt/controller/pool-controller/handler.go]: :)

**Pull request merged**